Writing tutorial documentation

genisplaja opened the pull request:
We are writing the tutorial documentation together with @PRamoneda. I am creating this PR already so we can hear your thoughts and opinions!!

magdalenafuentes reviewed (changes requested):
Thanks so much for this @genisplaja @PRamoneda, left some comments

genisplaja commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,171 @@ Tutorial
+
+.. code-block:: python
+
+    for key, track in orchset.load_tracks().items():
+        print(key, track.title, track.audio_path)
```
I think we can attach both! And then they can chose what works better for them.

magdalenafuentes replied:
I like the two examples! I feel it will also be nice to add an example when data is processed just to illustrate a pipeline with `mirdata`, maybe just recycle the example that used to be in the docs?

```
### Evaluate a melody extraction algorithm on Orchset
```python
import mir_eval
import mirdata
import numpy as np
import sox

def very_bad_melody_extractor(audio_path):
    duration = sox.file_info.duration(audio_path)
    time_stamps = np.arange(0, duration, 0.01)
    melody_f0 = np.random.uniform(low=80.0, high=800.0, size=time_stamps.shape)
    return time_stamps, melody_f0

# Evaluate on the full dataset
orchset_scores = {}
orchset = mirdata.Dataset('orchset')
orchset_data = orchset.load_tracks()
for track_id, track_data in orchset_data.items():
    est_times, est_freqs = very_bad_melody_extractor(track_data.audio_path_mono)

    ref_times = track_data.melody.times
    ref_freqs = track_data.melody.frequencies

    score = mir_eval.melody.evaluate(ref_times, ref_freqs, est_times, est_freqs)
    orchset_scores[track_id] = score

# Split the results by composer and by instrumentation
composer_scores = {}
strings_no_strings_scores = {True: {}, False: {}}
for track_id, track_data in orchset_data.items():
    if track_data.composer not in composer_scores.keys():
        composer_scores[track_data.composer] = {}

    composer_scores[track_data.composer][track_id] = orchset_scores[track_id]
    strings_no_strings_scores[track_data.contains_strings][track_id] = \
        orchset_scores[track_id]
```

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,240 @@ Tutorial
+        orchset = mirdata.Dataset('orchset')
+        orchset.download()  # Dataset is downloaded at user root folder
 
+Donwloading a dataset into a specified folder:
```
typo downloading

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,240 @@ Tutorial
+where the dataset is going to be stored and retrieved.
+
+Downloading a dataset into the default folder:
+    In this first example, ``data_home`` is not specified. Thus, ORCHSET will be downloaded and retrieved from mir_datasets
```
nit: mir_datasets -> `mir_datasets`

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,240 @@ Tutorial
+select which elements of the dataset they want to download. Each dataset has a ``REMOTES`` dictionary were all
+the available elements are listed.
+
+cante100 has different elements as seen in the REMOTES dictionary. Thus, we can specify which of these elements are
```
nit: REMOTES -> `REMOTES`

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,240 @@ Tutorial
+Using the method ``validate()`` we can check if the files in the local version are the same than the available canical version,
+and the files were downloaded correctly (none of them are corrupted).
 
+For big datasets: In future ``mirdata`` versions, a random validation will be included. This improvement will reduce validation time for very big datasets.
```
maybe include small code example with all files ok?

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,240 @@ Tutorial
+    example_melody = orchset.track(orchset_ids[0]).melody  # Get melody from first track in the index
+
+
+Annotations can also be accessed through ``load_someAnnotation()`` methods.
+
+.. code-block:: python
+
+    orchset_ids = orchset.track_ids  # Load list of track ids of the dataset
+    example_melody_path = orchset.track(orchset_ids[0]).melody_path  # Parsing melody annotation path
+
+    # Accessing to track melody annotation
+    example_melody = orchset.load_melody(example_melody_path)
+    print(example_melody.frequencies)
+    >>> array([  0.   ,   0.   ,   0.   , ..., 391.995, 391.995, 391.995])
+    print(example_melody.times)
+    >>> array([0.000e+00, 1.000e-02, 2.000e-02, ..., 1.244e+01, 1.245e+01, 1.246e+01])
```
I am wondering if this should go in the "working with remote datasets" section exclusively, since we want to favor  usage of the attributes directly? Let's discuss this

genisplaja replied:
Yes, IMO I think it's good to inform that the load_ functions can be used if needed, but that the main idea is to retrieve the annotations using the attributes of the dataset. Maybe we could add a minor point somewhere commenting the use of load_ functions and take it away from here.

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,240 @@ Tutorial
+``mirdata`` uses several diffent data classes to store annotations. These data classes are meant to standarize the organization for
+all the loaders, and keep compatibility with `JAMS <https://jams.readthedocs.io/en/stable/>`_ and `mir_eval <https://craffel.github.io/mir_eval/>`_.
+
+The list of available annotation classes are:
+
+- BeatData(times, positions)
+- SectionData(intervals, labels)
+- NoteData(intervals, notes, confidence)
+- ChordData(intervals, labels, confidence)
+- F0Data(times, frequencies, confidence)
+- MultiF0Data(times, frequency_list, confidence_list)
+- KeyData(intervals, keys)
+- LyricData(intervals, lyrics, pronunciations)
+- TempoData(intervals, value, confidence)
+- EventData(intervals, events)
```
For easier maintenance of the docs I'd simply include a link to the `annotations` sections in the docs, so in the future when we expand annotations we don't need to edit this one

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,240 @@ Tutorial
+In general, most datasets are a collection of tracks. Each track has an audio with its own annotations.
+
+With the ``load_tracks()`` method, all the tracks (so including their respective audio and annotations) can be loaded
+in a dictionary structure.
```
loaded as a dictionary with the ids as keys and tracks as items?

genisplaja replied:
Yes, that's better!

magdalenafuentes reviewed (commented):
Some more comments, most of them are nits!

genisplaja commented:
@magdalenafuentes about this comment... https://github.com/mir-dataset-loaders/mirdata/pull/378#discussion_r546708590 (sorry I am attaching it here but it won't allow me to comment in there). This example is already in `example.rst`. Wouldn't it be a bit redundant to add it here as well? We can reference it, so users can access to `example.rst` through a link and we avoid repetitions. What do you think?

magdalenafuentes commented:
Hey @genisplaja, about the example, the idea was to remove any examples from the repository and move it to the docs. If `exampels.rst` is somewhere in the docs that's ok (sorry if I'm not up-to-date, but I couldn't find it!) If it is not part of the docs, the tutorial could just include an example and everything would be in the same place. Let me know what do you think

rabitt commented:
> Hey @genisplaja, about the example, the idea was to remove any examples from the repository and move it to the docs. If `exampels.rst` is somewhere in the docs that's ok (sorry if I'm not up-to-date, but I couldn't find it!) If it is not part of the docs, the tutorial could just include an example and everything would be in the same place. Let me know what do you think

I removed examples.rst from the docs, but left it in the repo so the code could be moved into the tutorial! I'd definitely take everything from there, put it where it's relevant in tutorial.rst, and then remove examples.rst

genisplaja commented:
> > Hey @genisplaja, about the example, the idea was to remove any examples from the repository and move it to the docs. If `exampels.rst` is somewhere in the docs that's ok (sorry if I'm not up-to-date, but I couldn't find it!) If it is not part of the docs, the tutorial could just include an example and everything would be in the same place. Let me know what do you think
> 
> I removed examples.rst from the docs, but left it in the repo so the code could be moved into the tutorial! I'd definitely take everything from there, put it where it's relevant in tutorial.rst, and then remove examples.rst

Cool thanks! In fact, I already moved all the content in examples.rst to the tutorial.rst. I haven't copied and pasted the content from one file to the other but try to organize the code examples in examples.rst into de tutorial.rst structure. I think I already removed examples.rst in this PR!

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,395 @@ Tutorial
+select which elements of the dataset they want to download. Each dataset has a ``REMOTES`` dictionary were all
+the available elements are listed.
+
+cante100 has different elements as seen in the ``REMOTES`` dictionary. Thus, we can specify which of these elements are
```
nit: cante100 ->``cante100``

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,395 @@ Tutorial
+        ),
+    }
+
+An partial download example for cante100 dataset could be:
```
``cante100``

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,395 @@ Tutorial
+
+The list and descriptions of available annotation classes can be found `in this link <https://mirdata.readthedocs.io/en/latest/source/mirdata.html#module-mirdata.annotations>`_.
+
+**These classes are extendable in case a certain loader requires it.**
```
I would make this into a note for clarity

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,395 @@ Tutorial
+
+For the end user there is no difference between the remote and local indexes.
+
+However, to extend the library you have to add in utils.LargeData(...) the remote_index argument with a
```
extend the library you meant contribute?

magdalenafuentes replied:
utils.LargeData(...) -> ``utils.LargeData(...)``

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,395 @@ Tutorial
+For the end user there is no difference between the remote and local indexes.
+
+However, to extend the library you have to add in utils.LargeData(...) the remote_index argument with a
+download_utils.RemoteFileMetadata dictionary with the remote index information.
```
``download_utils.RemoteFileMetadata``

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,395 @@ Tutorial
+However, to extend the library you have to add in utils.LargeData(...) the remote_index argument with a
+download_utils.RemoteFileMetadata dictionary with the remote index information.
+
+DATA = utils.LargeData("acousticbrainz_genre_index.json", remote_index=REMOTE_INDEX)
```
Make this code block maybe?

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,395 @@ Tutorial
+    DATA = utils.LargeData("acousticbrainz_genre_index.json", remote_index=REMOTE_INDEX)
 
 
 Working with big datasets
 ^^^^^^^^^^^^^^^^^^^^^^^^^
 
+In the development of large datasets, it is advisable to create an index as small as possible to develop
+the new dataset and pass the tests.
```
This feels a comment for `contributing`, maybe move it there? The idea of having this section here was to explain the use of remote data, but it is addressed in `Accessing annotations remotely`

genisplaja replied:
Yes! That's is a good idea. I moved it into the testing section and I also added a reminder to keep the testing resources as light as possible!

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,395 @@ Tutorial
+    example_melody = orchset.track(orchset_ids[0]).melody  # Get melody from first track in the index
+
+
+Accessing annotations remotely
```
annotations and audio right?

magdalenafuentes commented on docs/source/tutorial.rst:
```diff
@@ -7,47 +7,395 @@ Tutorial
+Basic example: including mirdata in your pipeline
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If we wanted to use Orchset to evaluate the performance of a melody extraction algorithm
```
Orchset -> ``orchset``

magdalenafuentes reviewed (approved):
@genisplaja @PRamoneda this looks really great! Thanks :) 

I left minor comments, and if I remember well you were planing to collapso some code as the rest of the docs?

After that feel free to merge, great work!

magdalenafuentes commented:
Closes #364

genisplaja commented:
@rabitt, @magdalenafuentes I noticed that the `overview.rst` looks different between this version and the one in master. Which version should we keep? The one in this branch is more complete but the code snippets are not collapsible yet!

magdalenafuentes commented:
Mmmm.. I remember we discussed the overview with @rabitt but I don't remember in which branch though. Maybe @rabitt remembers? Otherwise I'd keep the most complete and add collapsable snippets

rabitt commented:
> @rabitt, @magdalenafuentes I noticed that the overview.rst looks different between this version and the one in master. Which version should we keep? The one in this branch is more complete but the code snippets are not collapsible yet!

Im a little confused - it looks like this PR doesn't change overview.rst at all. Am i missing something?

genisplaja commented:
@rabitt @magdalenafuentes My bad!! I got confused between branches, sorry! So I guess this PR is ready!

rabitt commented:
Amazing! I'm merging. Thanks for all the work!
